mailbox/backup/src/main/java/org/apache/james/mailbox/backup/ZipMailArchiveRestorer.java

```diff
@@ -0,0 +1,101 @@
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.mailbox.backup;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.james.core.User;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageManager;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.util.OptionalUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
+public class ZipMailArchiveRestorer implements MailArchiveRestorer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ZipMailArchiveRestorer.class);
+
+    private final MailboxManager mailboxManager;
+    private final MailArchivesLoader archiveLoader;
+
+    public ZipMailArchiveRestorer(MailboxManager mailboxManager, MailArchivesLoader archiveLoader) {
+        this.mailboxManager = mailboxManager;
+        this.archiveLoader = archiveLoader;
+    }
+
+    public void restore(User user, InputStream source) throws MailboxException, IOException {
+        MailboxSession session = mailboxManager.createSystemSession(user.asString());
+        restoreEntries(source, session);
+    }
+
+    private void restoreEntries(InputStream source, MailboxSession session) throws IOException {
+        try (MailArchiveIterator archiveIterator = archiveLoader.load(source)) {
+            List<MailboxWithAnnotationsArchiveEntry> mailboxes = readMailboxes(archiveIterator);
+            restoreMailboxes(session, mailboxes);
```

Comment:
seems a bad idea for me: you have an iterator, but you transform it into a list. So you are loading all the archive into memory.
Here it's only mailboxes, so it's not a big deal, but do you plan to fix it later?

Reply:
It's done only for the mailboxes. It will not be the case for the messages. 
This design choice was pushed by the team who wanted at the time to restore the mailboxes, then messages.

But indeed when we will restore the messages, it makes sense not to use a Map of Mailbox by the id of the corresponding zipped mailbox. But to iterate and for each mailbox restore its messages at the same time. It's what was originally done.